### PR TITLE
Expressions: Optimization for string expressions on the __time column.

### PR DESCRIPTION
**Pull request by gianm**
_WIP: Still need to run benchmarks_

### CODE REVIEW

**Comment by drcrallen** [comment]
@gianm can you color this with a couple of use cases that highlight the need for optimization?

**Comment by gianm** [comment]
@drcrallen,

> @gianm can you color this with a couple of use cases that highlight the need for optimization?

It's an expression oriented equivalent of the optimization that SingleScanTimeDimSelector does for extractionFns on the __time column. It helps because it takes advantage of the fact that the time column tends to have a lot of runs, and with the optimization, we only compute the function once per distinct value rather than computing it repeatedly.

**Comment by leventov** [comment]
I think it should be precomputed for column and stored along with the column, if such optimization makes sense for a column. Along with min, max, average for columns, and other types of O(1) information that could be useful during querying. It should be accounted here: https://github.com/druid-io/druid/projects/1

File: benchmarks/src/main/java/io/druid/benchmark/ExpressionSelectorBenchmark.java [hunk 1]
```diff
@@ -124,6 +124,80 @@ public void tearDown() throws Exception
...
+    final Sequence<Cursor> cursors = new QueryableIndexStorageAdapter(index).makeCursors(
+        null,
+        index.getDataInterval(),
+        VirtualColumns.EMPTY,
+        Granularities.ALL,
+        false,
+        null
+    );
+
+    final List<?> results = Sequences.toList(
```
[the remark this answers is not in the document]
Comment (leventov) [reply benchmarks/src/main/java/io/druid/benchmark/ExpressionSelectorBenchmark.java]: However I did it myself in #5175.

**Comment by fjy** [comment]
@gianm we should finish this off

**Comment by gianm** [comment]
It turns out this patch is not necessary, since #5048 accomplished the same goal. String expressions on the __time column work by first building a base primitive selector, and #5048 optimizes those. That optimization ends up being inherited by the dimension selector.

#6599 extends the optimization to apply to Long columns that are not __time.

**Pull request closed without merging**